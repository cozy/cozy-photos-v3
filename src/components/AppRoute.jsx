import React from 'react'
import { Route, Redirect } from 'react-router'

import Layout from './Layout'
import Timeline from '../containers/Timeline'
import AlbumsView from '../containers/AlbumsView'
import AlbumPhotos from '../containers/AlbumPhotos'
import Viewer from '../components/Viewer'
import { PhotosPicker } from '../ducks/albums'

export const ComingSoon = () => (<p style='margin-left: 2em'>Coming soon!</p>)

const AppRoute = (
  <Route component={Layout}>
    <Route path='photos' component={Timeline}>
      <Route path=':photoId' component={Viewer} />
    </Route>
    <Route path='albums' component={AlbumsView}>
      <Route path='new' component={PhotosPicker} />
      <Route path=':albumId' component={AlbumPhotos}>
        <Route path='edit' component={PhotosPicker} />
        <Route path=':photoId' component={Viewer} />
      </Route>
    </Route>
    <Redirect from='/*' to='photos' />
  </Route>
)

export default AppRoute
