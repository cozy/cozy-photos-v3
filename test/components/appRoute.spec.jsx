'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import AppRoute, { ComingSoon } from '../../src/components/AppRoute'

describe('AppRoute component', () => {
  it('should render all the routes correctly with their components', () => {
    expect(AppRoute.props.children).toBeInstanceOf(Array)

    /* test all routes and nested routes */
    // components connected with Redux will be displayed as 'Connect(componentName)'
    // ComingSoon is not a component so -> not defined
    // translated component will be display as '_translate'
    const expectedStructure = {
      photos: {
        path: 'photos',
        component: 'Connect(Timeline)',
        routes: { // nested
          ':photoId': {
            path: ':photoId',
            component: 'Connect(withRouter(Viewer))'
          }
        }
      },
      albums: {
        path: 'albums'
      },
      shared: {
        path: 'shared'
      },
      trash: {
        path: 'trash'
      }
    }

    const checkRoute = (element, expectedStructure) => {
      if (element.type.displayName === 'Route') { // if Route component
        expect(element.props.path).toBe(
          expectedStructure[element.props.path].path
        )

        if (element.props.component) { // if there is a component
          if (element.props.component.displayName) { // connected component
            expect(element.props.component.displayName).toBe(
              expectedStructure[element.props.path].component
            )
          } else if (element.props.component.name) {
            expect(element.props.component.name).toBe(
              expectedStructure[element.props.path].component
            )
          }
        }

        if (element.props.children) { // if subroutes
          if (Array.isArray(element.props.children)) {
            element.props.children.forEach((element) => {
              checkRoute(element, expectedStructure[element.props.path].routes)
            })
          } else {
            checkRoute(
              element.props.children,
              expectedStructure[element.props.path].routes
            )
          }
        }
      }
    }

    AppRoute.props.children.forEach((element) => {
      checkRoute(element, expectedStructure)
    })
  })

  it('should temporary use a ComingSoon function as component for routes not available yet', () => {
    const comingSoonTextElement = ComingSoon()
    expect(comingSoonTextElement).toMatchSnapshot()
  })
})
