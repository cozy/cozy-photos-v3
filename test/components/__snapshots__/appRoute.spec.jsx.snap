// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppRoute component should temporary use a ComingSoon function as component for routes not available yet 1`] = `
<p
  style="margin-left: 2em"
>
  Coming soon!
</p>
`;
